### ISSUE TITLE
v2: Add support for SQUASH_ACTIONS to AnsibleModule

### ISSUE THREAD

**Issue by kustodian** [ansible/ansible#11383]
In v1 and now I see from the code in v2 as well, for Ansible to squash `with_*` items into a single item and pass it to the module at once, you needed to modify the core to add the module into a list of modules which support this. Currently you need to edit `lib/ansible/executor/task_executor.py`

```
SQUASH_ACTIONS = frozenset(('apt', 'yum', 'pkgng', 'zypper', 'dnf'))
```

Also this doesn't pass a list of those items to the module, but joins those items with `join(',')`. Would it be possible that v2 adds support in the `AnsibleModule` class so you can pass an argument e.g. `squash_actions=True` which will automatically pass all the `with_*` items as a list of items to the module. This should make the execution of a lot modules (if they add support) a lot faster, but it will also make it easier to add this support to future (and old) modules since you don't have to make a separate pull request to the core to enable this for a module.

**Comment by bcoca** [comment]
I had previously attempted to make this configurable in v1 (moving the module string to ansible.cfg), I need to look closer at the current code to see if it would be simpler to implement as you suggest, also to be smarter about how the list is composed and not assume that name="{{item}}" to allow for string composition.

**Comment by sivel** [comment]
I had looked at this at one point too.  Part of the problem is that with_items processing happens in the executor, and with_items is not exposed to the module.

The problem is execution order.  The `AnsibleModule` class is instantiated after the executor does it's thing, so attributes applied in the module are not accessible at that time.

Maybe adding something similar to the `WANT_JSON` tag in a module could expose this early enough to the executor.  Such as a `SQUASH_ACTIONS` tag.  This would probably involve a fair amount of moving of code, but if the module is to define that it wants squashed actions, I can't think of may other options.

**Comment by kustodian** [comment]
I remember talking to @jimi-c about this during AnsibleFest in London, and he said it should be possible to do, but I forgot to create an issue about it so that he remembers to check it out.

@sivel if it can be done by adding a tag as you suggested, is totally ok if you ask me, even if it feels like a hack, since it would still be a smaller hack than how it currently works.

**Comment by jimi-c** [comment]
@kustodian I believe I meant more along the lines of what @bcoca was describing, ie. making it more configurable rather than hard-coded. Doing this in TaskExecutor is the best way, however making it a bit more flexible would be nice. The main thing people want to do is this:

<pre>
name=something-{{item}}
</pre>

...which has never worked, due to the fact that we template just the variable portion and not the name field.

**Comment by kustodian** [comment]
I didn't look into the Ansible core that much, I was mostly working on modules, so I'm of no help here, but would it be possible to do what you described and are you guys willing to do it? :)

**Comment by jimi-c** [comment]
I have a patch that seems to make this work:

<pre>
diff --git a/lib/ansible/executor/task_executor.py b/lib/ansible/executor/task_executor.py
index ddd557f..8405389 100644
--- a/lib/ansible/executor/task_executor.py
+++ b/lib/ansible/executor/task_executor.py
@@ -186,8 +186,14 @@ class TaskExecutor:
                 variables['item'] = item
                 templar = Templar(loader=self._loader, shared_loader_obj=self._shared_loader_obj, variables=variables)
                 if self._task.evaluate_conditional(templar, variables):
-                    final_items.append(item)
-            return [",".join(final_items)]
+                    if templar._contains_vars(self._task.args['name']):
+                        new_item = templar.template(self._task.args['name'])
+                        final_items.append(new_item)
+                    else:
+                        final_items.append(item)
+            joined_items = ",".join(final_items)
+            self._task.args['name'] = joined_items
+            return [joined_items]
         else:
             return items
 
</pre>

**Comment by kustodian** [comment]
So that should make `name=something-{{item}}` work :)

Now how to make this configurable from the module itself, since there will be modules which don't use `name` argument, or they have more arguments, for example:

```
- name: Set up authorized_keys for user1
  authorized_key: user=user1 key="{{ item }}"
  with_file:
    - key1
    - key2
```

or:

```
- name: Configure sysctl.conf
  sysctl: name={{ item.key }} value={{ item.value }} state=present
  with_dict: common_sysctl_entries
```

In the first situation `key` argument would need to be passed as a list, while in the second two arguments `name` and `value`.

**Comment by jimi-c** [comment]
Doing this on a per-module basis would be way more tricky, as they would have to all be updated to have the name parameter accept a list, and to act accordingly if there was more than one item in that list. That would take quite a bit more work.

**Comment by jimi-c** [comment]
Also, this would only work if the item variable was in the name, when using a dictionary or multi-dimensional list this would be much more complicated, as we'd have to look through all module params and pass the items as lists there, and that's not something we're going to do.

**Comment by kustodian** [comment]
Actually you wouldn't need to go over all modules (only those 5 modules which currently support this), you would just need to add support for it and modules can be updated later on (if they gain anything from it). It would just need to be possible to define what module arguments supports squashing (since currently it is hardcoded for the `name` argument.

My idea was to define in the module definition (however possible) a list of arguments which support lists from `with_*`. For example it could be called `SQUASH_ARGS=['name', 'value']` in case of `sysctl` module, and if `SQUASH_ARGS` is not defined, everything works like before.

**Comment by kustodian** [comment]
@bcoca how can we use this in a module?

**Comment by bcoca** [comment]
just add module name to the configuration and have module check if what it gets is a list (or always make it a list even if just a string) and then loop over it.

**Comment by kustodian** [comment]
So for the module to support this by default, you still need to edit core? Because this is taken care of in the executor, I guess it is hard to make it configurable in the module itself, so this solution at least makes not look like a hack in the core.

**Comment by bcoca** [comment]
no, that used to be the case, now its configurable, so you just need to edit ansible.cfg (squash_actions)

**Comment by bcoca** [comment]
@kustodian sorry, i thought i had merged my code, what i stated above is only true once PR #11880 is merged

**Comment by kustodian** [comment]
Of course, I knew it meant once you merged your changes. What I wanted to say is if you want to make it a default for a module, you need to change core. At least it's easy to find, instead of digging the code.

**Comment by bcoca** [comment]
you would have to change all modules to support it first.

**Comment by kustodian** [comment]
I was thinking about setting a flag in the module itself so that the executor knows the module supports squashable items, but that's currently not possible to do.

**Comment by jctanner** [comment]
resolved_by_pr https://github.com/ansible/ansible/pull/11880
